tests/test_project.py

```diff
@@ -2317,6 +2319,184 @@ def test_main_submit(self, monkeypatch):
+    project_class = define_hooks_test_project._HooksTestProject
+    operation_name = "base"
+    entrypoint = dict(
+        path=os.path.realpath(
+            os.path.join(os.path.dirname(__file__), "define_hooks_test_project.py")
+        )
+    )
+
+    @staticmethod
+    def _get_job_doc_key(job, operation_name):
+        return lambda key: job.doc.get(f"{operation_name}_{key}")
+
+    @pytest.fixture(params=["base"])
+    def operation_name(self, request):
+        return request.param
+
+    def mock_project(self):
+        project = self.project_class.get_project(root=self._tmp_dir.name)
+        project.open_job(dict(raise_exception=False)).init()
+        project.open_job(dict(raise_exception=True)).init()
+        project = project.get_project(root=self._tmp_dir.name)
+        project._entrypoint = self.entrypoint
+        return project
+
+    def call_subcmd(self, subcmd, stderr=subprocess.DEVNULL):
+        # Bypass raising the error/checking output since it interferes with hook.on_fail
+        fn_script = self.entrypoint["path"]
+        _cmd = f"python {fn_script} {subcmd}"
+        with add_path_to_environment_pythonpath(os.path.abspath(self.cwd)):
+            with switch_to_directory(self.project.root_directory()):
+                return subprocess.run(_cmd.split())
+
+    @pytest.fixture(scope="function")
+    def project(self):
+        return self.mock_project()
+
+    @pytest.fixture(params=[True, False], ids=["raise_exception", "no_exception"])
+    def job(self, request, project):
+        return project.open_job(dict(raise_exception=request.param))
+
+    def test_start_and_finish(self, project, job, operation_name):
+        get_job_doc_value = self._get_job_doc_key(job, operation_name)
+
+        assert get_job_doc_value(self.keys[0]) is None
+        assert get_job_doc_value(self.keys[1]) is None
+
+        self.call_subcmd(f"run -o {operation_name} -j {job.id}")
+
+        assert get_job_doc_value(self.keys[0])
+        assert get_job_doc_value(self.keys[1])
+
+    def test_success(self, project, job, operation_name):
+        get_job_doc_value = self._get_job_doc_key(job, operation_name)
+
+        assert get_job_doc_value(self.keys[2]) is None
+
+        self.call_subcmd(f"run -o {operation_name} -j {job.id}")
+
+        if job.sp.raise_exception:
+            assert not get_job_doc_value(self.keys[2])
+        else:
+            assert get_job_doc_value(self.keys[2])
+
+    def test_fail(self, project, job, operation_name):
+        get_job_doc_value = self._get_job_doc_key(job, operation_name)
+
+        assert get_job_doc_value(self.keys[3]) is None
+
+        self.call_subcmd(f"run -o {operation_name} -j {job.id}")
+
+        if job.sp.raise_exception:
+            assert get_job_doc_value(self.keys[3])[0]
+            assert get_job_doc_value(self.keys[3])[1] == self.error_message
+        else:
+            assert get_job_doc_value(self.keys[3]) is None
+
+
+class TestHooksInstallBase(TestHooksBase):
```

Comment:
I am confused by these tests. I am likely just missing it, but shouldn't the project level hooks ensure that the job document is edited for all operations not just those with hook decorators?

Reply:
These test should check operations with and without decorators in the functions inherited from `TestHooksBase`, where `test_start_finish` will test job document keys "installed_start" and "install_finish", `test_success` will test "installed_success", and `test_fail` will test "installed_fail" for operations with and without decorators. 

Tests with the name `test_decorated_*` are only run on operations with decorators and check for changes to the job document from those decorators (i.e. job document keys "start", "finish" instead of "installed_start" or "installed_finish").

Reply:
I guess what I mean is why don't we test that `base_no_decorators` has the keys `install_finish`, etc. written?

Reply:
Oh  the `install_*` for `base_no_decorators` is tested! The operations tested in `TestHooksInstallBase` are paramaterized like this:
https://github.com/glotzerlab/signac-flow/blob/7ea174f5505c3d55dfbe1cc6269920426e535012/tests/test_project.py#L2410-L2412
While in `TestHooksInstallCmd`, `operation_name` is paramaterized as `base_cmd` and `base_cmd_no_decorators`.

The test functions inherited from `TestHooksBase` all accept `operation_name` as a parameter and run operations as such:
https://github.com/glotzerlab/signac-flow/blob/7ea174f5505c3d55dfbe1cc6269920426e535012/tests/test_project.py#L2371

Does that make sense?

Reply:
So, I understand now how the test work, but @klywang and I discussed ways to improve test clarity.

1. comment what class level variables/fixtures like `keys` and `operation_name` do.
2. Remove tests from `TestHooksInstallBase`, and move tests that change that `base_no_decorators` does not write `start`, `finish`, etc. job document keys in the `TestHooksBase` class. You may want to make the no decorator operation name a variable to allow you to reuse this for the cmd tests.
3. Make the aforementioned test only test that `base_no_decorators` does not have the decorator level hooks.
4. Rename `keys` to `job_doc_keys` and make it a dictionary with keys `"start", "finish", "success", "fail"`.